### PR TITLE
Allow ICNS save on all operating systems

### PR DESCRIPTION
**Pull request by newpanjing**
Changes proposed in this pull request:

 * Save as icns to support all operating systems
 * Removing the restriction that you can only use macOS

### CODE REVIEW

**Comment by radarhere** [comment]
Thanks very much. Is there documentation that you followed when writing this?

File: src/PIL/IcnsImagePlugin.py [hunk 1]
```diff
@@ -6,22 +6,21 @@
 #
 # history:
 # 2004-10-09 fl   Turned into a PIL plugin; removed 2.3 dependencies.
+# 2020-04-04 Save in icns format to support all operating systems.
```
Comment (radarhere) [inline src/PIL/IcnsImagePlugin.py:9]:
```suggestion
# 2020-04-04 Allow saving on all operating systems.
```

File: src/PIL/IcnsImagePlugin.py [hunk 2]
```diff
@@ -302,65 +300,70 @@ def load(self):
...
+
+MAGIC = _to_int("icns")
+TOC = "TOC "
+
+
 def _save(im, fp, filename):
     """
     Saves the image as a series of PNG files,
     that are then converted to a .icns file
-    using the macOS command line utility 'iconutil'.
```
Comment (hugovk) [inline src/PIL/IcnsImagePlugin.py:316]:
```suggestion
```

File: src/PIL/IcnsImagePlugin.py [hunk 3]
```diff
@@ -302,65 +300,70 @@ def load(self):
...
 
-    macOS only.
+    Support for arbitrary systems
```
Comment (hugovk) [inline src/PIL/IcnsImagePlugin.py:317]: We can skip this, it's a bit redundant
```suggestion
```

File: src/PIL/IcnsImagePlugin.py [hunk 4]
```diff
@@ -302,65 +300,70 @@ def load(self):
...
-            im_w2 = provided_images.get(w * 2, im.resize((w * 2, w * 2), Image.LANCZOS))
-            im_w2.save(second_path)
-            last_w = w * 2
-
-        # iconutil -c icns -o {} {}
-
-        convert_cmd = ["iconutil", "-c", "icns", "-o", filename, iconset]
-        convert_proc = subprocess.Popen(
-            convert_cmd, stdout=subprocess.PIPE, stderr=subprocess.DEVNULL
+    # size
```
Comment (hugovk) [inline src/PIL/IcnsImagePlugin.py:322]: For consistency with `# Header`, etc
```suggestion
    # Size
```

File: docs/handbook/image-file-formats.rst [hunk 5]
```diff
@@ -200,12 +200,16 @@ attributes before loading the file::
...
-Pillow reads and (macOS only) writes macOS ``.icns`` files.  By default, the
+Pillow reads and writes macOS ``.icns`` files.  By default, the
 largest available icon is read, though you can override this by setting the
 :py:attr:`~PIL.Image.Image.size` property before calling
 :py:meth:`~PIL.Image.Image.load`.  The :py:meth:`~PIL.Image.Image.open` method
 sets the following :py:attr:`~PIL.Image.Image.info` property:
 
+.. note::
+
+    Prior to version 7.2.0, Pillow could only write ICNS files on macOS.
```
Comment (radarhere) [inline docs/handbook/image-file-formats.rst:211]:
```suggestion
    Prior to version 8.1.0, Pillow could only write ICNS files on macOS.
```

**Comment by BoboTiG** [comment]
That would be a good thing! What is the status of the PR @newpanjing ?

**Review by hugovk — APPROVED** [review]

File: src/PIL/IcnsImagePlugin.py [hunk 6]
```diff
@@ -302,73 +300,68 @@ def load(self):
...
+    # Data
+    for index, e in enumerate(entries):
+        fp.write(struct.pack("<i", _to_int(e["type"]))[::-1])
+        fp.write(struct.pack("<i", HEADERSIZE + e["size"])[::-1])
+        fp.write(e["stream"].getvalue())
 
-        if fp_only:
-            with open(filename, "rb") as f:
-                fp.write(f.read())
+    fp.flush()
```
Comment (radarhere) [inline src/PIL/IcnsImagePlugin.py:355]:
```suggestion
    if hasattr(fp, "flush"):
        fp.flush()
```
Mimicking https://github.com/python-pillow/Pillow/pull/4526/files#diff-bb9af648f9d6055f580440e268d715dae9a22a2c4bfd47afd18bd834c05e96d2R317-R318

**Review by hugovk — APPROVED** [review]

**Pull request merged**